### PR TITLE
move external url controller to commit object

### PR DESCRIPTION
**Pull request by acosferreira**
# Description

move the external url controller to commit object to avoid errors on resume images

Fixes # ([THEEDGE-2917](https://issues.redhat.com/browse/THEEDGE-2917))

## Type of change

What is it?

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update
- [ ] Tests update
- [ ] Refactor

# Checklist:

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [X] I run `make pre-commit` to check fmt/vet/lint/test-no-fdo

### CODE REVIEW

File: pkg/services/repobuilder.go [hunk 1]
```diff
@@ -280,7 +285,11 @@ func (rb *RepoBuilder) DownloadVersionRepo(c *models.Commit, dest string, extern
 		}
 	}
 	rb.log.Info("Download finished")
-
+	c.ExternalURL = true
```
Comment (ldjebran) [inline pkg/services/repobuilder.go:288]: we should set this value as true the first time when we get the url from image builder eg in the client after line 
https://github.com/RedHatInsights/edge-api/blob/main/pkg/clients/imagebuilder/client.go#L392 

```go
                image.Commit.Status = models.ImageStatusSuccess
		image.Commit.ImageBuildTarURL = cs.ImageStatus.UploadStatus.Options.URL
		image.Commit.ExternalURL = true
```

File: pkg/services/repobuilder.go [hunk 2]
```diff
@@ -81,7 +81,12 @@ func (rb *RepoBuilder) BuildUpdateRepo(id uint) (*models.UpdateTransaction, erro
 	if err != nil {
 		return nil, err
 	}
-	tarFileName, err := rb.DownloadVersionRepo(update.Commit, path, false)
+	update.Commit.ExternalURL = false
```
Comment (ldjebran) [inline pkg/services/repobuilder.go:84]: we should set this value to false at the place where we set the internal url  eg after line 
https://github.com/RedHatInsights/edge-api/blob/main/pkg/services/repobuilder.go#L320

```go
        c.ImageBuildTarURL = repoTarURL   // line 320
        c.ExternalURL = false
	result := db.DB.Save(c)
```

**Review by ldjebran — APPROVED** [review]
/lgtm

File: pkg/clients/imagebuilder/client.go [hunk 3]
```diff
@@ -318,10 +318,12 @@ func (c *Client) ComposeInstaller(image *models.Image) (*models.Image, error) {
 	if err != nil {
 		image.Installer.Status = models.ImageStatusError
 		image.Status = models.ImageStatusError
+		image.Commit.ExternalURL = false
```
Comment (omaciel) [inline pkg/clients/imagebuilder/client.go:321]: This line is repeated twice inside this conditional statement. Is this intentional?

**Review by omaciel — CHANGES_REQUESTED** [review]
We don't have any tests to check your changes? I think we need one for the conditional statement I mentioned in `/pkg/clients/imagebuilder/client.go`.

File: pkg/clients/imagebuilder/client.go (hunk 3 above)
Comment (acosferreira) [reply pkg/clients/imagebuilder/client.go:321]: yeah, we check the status of commit while is building I want to keep as false, if an error occurs is also false.
 maybe i can move it for outside of validation and keep in one line ;)

**Review by omaciel — DISMISSED** [review]
/lgtm

**Review by thom-at-redhat — DISMISSED** [review]

File: pkg/services/repobuilder_test.go [hunk 4]
```diff
@@ -57,6 +64,24 @@ var _ = Describe("RepoBuilder Service Test", func() {
...
 	})
+
+	Describe("#DownloadVersionRepo", func() {
+		When("is valid", func() {
+			It("should download the repo", func() {
+				filePath := "/tmp/download/"
+				fmt.Print(filePath)
+				commit := &models.Commit{ExternalURL: true, ImageBuildTarURL: "https://repos.fedorapeople.org/pulp/pulp/demo_repos/zoo/bear-4.1-1.noarch.rpm"}
+				// mockFilesService.EXPECT().GetDownloader().Return(nil)
+				// mockDownloaderService.EXPECT().DownloadToPath(commit.ImageBuildTarURL, "repo.tar").Return(nil)
```
Comment (de1987) [inline pkg/services/repobuilder_test.go:75]: Could you remove those?

File: pkg/services/repobuilder.go [hunk 5]
```diff
@@ -272,9 +273,8 @@ func (rb *RepoBuilder) DownloadVersionRepo(c *models.Commit, dest string, extern
 		}
 	} else {
 		rb.log.WithFields(log.Fields{"filepath": tarFileName, "imageBuildTarURL": c.ImageBuildTarURL}).Debug("Downloading tar file")
-
-		filesService := NewFilesService(rb.log.WithField("url", c.ImageBuildTarURL))
-		if err := filesService.GetDownloader().DownloadToPath(c.ImageBuildTarURL, tarFileName); err != nil {
+		downloader := rb.FilesService.GetDownloader()
+		if err := downloader.DownloadToPath(c.ImageBuildTarURL, tarFileName); err != nil {
```
Comment (de1987) [inline pkg/services/repobuilder.go:277]: @ldjebran we replaced the `FilesService` init with the one that exists inside of `RepoBuilderService`
Is there any concern about it?

Comment (ldjebran) [reply pkg/services/repobuilder.go:277]: @de1987 I found this implementation more logical , as Ana want to create a mock for testing, there is no functionality changes here , so it lgtm

File: pkg/services/repobuilder_test.go [hunk 6]
```diff
@@ -57,6 +59,60 @@ var _ = Describe("RepoBuilder Service Test", func() {
...
+				Service:      services.NewService(ctx, log.NewEntry(log.StandardLogger())),
+				FilesService: mockFilesService,
+				Log:          &log.Entry{},
+			}
+			defer ctrl.Finish()
+
+		})
+		When("is valid internal url", func() {
+			It("should download the repo", func() {
+				fileDest := "/tmp/download/"
```
Comment (ldjebran) [inline pkg/services/repobuilder_test.go:86]: may be this should go to setup section lines After subscribe

File: pkg/services/repobuilder_test.go [hunk 7]
```diff
@@ -57,6 +59,60 @@ var _ = Describe("RepoBuilder Service Test", func() {
...
+
+		When("is valid external url", func() {
+			It("should download the repo", func() {
+				fileDest := "/tmp/download/"
+				fileName := "repo.tar"
+				commit := &models.Commit{ExternalURL: true, ImageBuildTarURL: fileURL}
+				n, err := downloadservice.DownloadVersionRepo(commit, fileDest)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(n).To(Equal(fmt.Sprintf("%v%v", fileDest, fileName)))
+				os.RemoveAll(fileDest)
```
Comment (ldjebran) [inline pkg/services/repobuilder_test.go:111]: may be this should go the AfterEach section

**Comment by ldjebran** [comment]
remove do not merge as automation is ok now

**Review by ldjebran — DISMISSED** [review]
/lgtm

**Review by de1987 — DISMISSED** [review]

**Review by de1987 — APPROVED** [review]

**Review by thom-at-redhat — APPROVED** [review]

**Pull request merged**